piksi_tools/console/settings_view.py

```diff
@@ -505,41 +475,82 @@ def display_ins_output_hint(self):
             callback=self._save_and_reset)
 
         confirm_prompt2.text = "\n\n" \
-                               "    In order for the \"Ins Output Mode\" setting to take effect, it is necessary to save the    \n" \
-                               "    current settings to device flash and then power cycle your device.    \n\n" \
-                               "    Choose \"OK\" to immediately save settings to device flash and send the software reset command.    \n" \
-                               "    The software reset will temporarily interrupt the console's connection to the device but it   \n" \
-                               "    will recover on its own.    \n\n"
+            "    In order for the \"Ins Output Mode\" setting to take effect, it is necessary to save the    \n" \
+            "    current settings to device flash and then power cycle your device.    \n\n" \
+            "    Choose \"OK\" to immediately save settings to device flash and send the software reset command.    \n" \
+            "    The software reset will temporarily interrupt the console's connection to the device but it   \n" \
+            "    will recover on its own.    \n\n"
 
         confirm_prompt2.run(block=False)
 
-    def _send_pending_settings_by_index(self):
-        for eachindex in self.pending_settings:
-            self.link(MsgSettingsReadByIndexReq(index=eachindex))
-
-    def _restart_retry_thread(self):
-        if self.retry_pending_read_index_thread:
-            self.retry_pending_read_index_thread.stop()
-        self.retry_pending_read_index_thread = TimedDelayStoppableThread(
-            SETTINGS_RETRY_TIMEOUT,
-            target=self._send_pending_settings_by_index, args=[])
-        self.retry_pending_read_index_thread.start()
-
-    def _settings_read_by_index(self):
-        self.enumindex = 0          # next index to ask for
-        self.pending_settings = []  # list of settings idices we've asked for
-        self.ordering_counter = 0   # helps make deterministic order of settings
-        self.setup_pending = True   # guards against receipt of multiple "done" msgs
-        # queue up BATCH_WINDOW settings indices to read
-        self.pending_settings = range(self.enumindex, self.enumindex + BATCH_WINDOW)
-        self.enumindex += BATCH_WINDOW
-        self._send_pending_settings_by_index()
-        # start a thread that will resend any read indexes that haven't come
-        self._restart_retry_thread()
+    def _settings_unconfirm_all(self):
+        for section in self.settings.keys():
+            for name in self.settings[section].keys():
+                self.settings[section][name].confirmed_set = False
+
+    def _settings_read_all(self):
+        self._settings_unconfirm_all()
+        threading.Thread(target=self._settings_read_all_thd).start()
```

Comment:
Let's make a task queue for these operations so we can be more disciplined about when threads are created.  If we need more than one thing to run at the same time we can use a thread pool to execute the tasks.

Reply:
Could you elaborate what exactly you mean by task queue?

Reply:
A thread (or pool of threads) that waits on a shared queue and executes tasks (functions) as they are enqueued.  Which (coincidentally?) looks like exactly what you've already implemented! 😄